### PR TITLE
fix: s3 source cannot read incremental files

### PR DESCRIPTION
**Pull request by tabVersion**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

prev fix does not handle the case right https://github.com/risingwavelabs/risingwave/pull/17702

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/connector/src/source/reader/reader.rs
Comment (fuyufjh) [inline src/connector/src/source/reader/reader.rs:209]: What's this for? Please add a comment

1 second sounds trivial, but considering our barrier interval is just 1 second, we need to be cautious.

Comment (tabVersion) [reply src/connector/src/source/reader/reader.rs:209]: to avoid unnecessary requests to s3 service. If using s3 source, users don't expect new files to be discovered within seconds. 
The sleep here does not block the barrier because it occurs in the file list stream, before merging with the barrier stream.

Comment (wcy-fdu) [reply src/connector/src/source/reader/reader.rs:209]: Shall we use `interval.tick()` instead of `sleep` to handle the incremental files? Thus we can make the update interval longer, like 1min(IIRC for origin s3, it's 10 mins).
Can can we make the update interval configured by user?

Comment (fuyufjh) [reply src/connector/src/source/reader/reader.rs:209]: I see. Agree with @wcy-fdu it can be longer and configurable.

**Comment by hzxa21** [comment]
Maybe related: #17991

File: e2e_test/s3/fs_source_v2_new_file.py
Comment (xxchan) [inline e2e_test/s3/fs_source_v2_new_file.py:75]: Can we make this test more robust? e.g., sth like: 1. upload some files. 2. check. 3. upload another batch of files. 4. check again

Comment (tabVersion) [reply e2e_test/s3/fs_source_v2_new_file.py:75]: I think it does the same thing

first check https://github.com/risingwavelabs/risingwave/blob/d5f503e15663beab4e73005a47ef08bfca67a740/e2e_test/s3/fs_source_v2_new_file.py#L70 

check after upload
https://github.com/risingwavelabs/risingwave/blob/d5f503e15663beab4e73005a47ef08bfca67a740/e2e_test/s3/fs_source_v2_new_file.py#L79

Comment (xxchan) [reply e2e_test/s3/fs_source_v2_new_file.py:75]: Then why it didn't catch the bug before? :eyes:

Comment (tabVersion) [reply e2e_test/s3/fs_source_v2_new_file.py:75]: I am surprised, too. 😄

Comment (xxchan) [reply e2e_test/s3/fs_source_v2_new_file.py:75]: So can we make this test more robust like doing **interleaving** upload and check? If a single `sleep` can affect the correctness, it seems too fragile.

**Comment by wcy-fdu** [comment]
New update:
1. Fixed the issue of being unable to read stock files
2. Temporarily set the sleep interval to 1 minute, and try to make it a configuration item later. (open an issue to track https://github.com/risingwavelabs/risingwave/issues/18123)

File: e2e_test/s3/fs_source_v2_new_file.py
Comment (xxchan) [reply e2e_test/s3/fs_source_v2_new_file.py:75]: Sorry, I misunderstood it. `do_test` is the first check, and `check_for_new_files` is the second check. I just looked for the usage of `check_for_new_files` and thought it's the only check.

**Review by wcy-fdu — APPROVED** [review]

**Comment by wcy-fdu** [comment]
Since I also modified this PR, this PR needs approval from another reviewer( @xxchan @fuyufjh ). Basically
, this PR has fixed the problem of reading newly added files. Currently, json and csv encode have been fixed. There is another problem with parquet encode, which will be fixed in the another [PR](https://github.com/risingwavelabs/risingwave/pull/18149).

**Review by xxchan — CHANGES_REQUESTED** [review]
Please fix test and CI

File: src/connector/src/source/reader/reader.rs
Comment (tabVersion) [reply src/connector/src/source/reader/reader.rs]: https://github.com/risingwavelabs/risingwave/pull/18184
The time interval is now configurable and will be carried into main in this pr

**Comment by wcy-fdu** [comment]
This test(`S3 source and sink on parquet file`) is a flacky test. It will also fail in other PRs, regardless of the changes made in this PR.

**Review by xxchan — APPROVED** [review]

File: src/stream/src/executor/source/list_executor.rs
Comment (xxchan) [inline src/stream/src/executor/source/list_executor.rs:81]: Why do we change to 1 file 1 chunk here? :eyes:
The name `chunked_paginate_stream` is confusing

Comment (tabVersion) [reply src/stream/src/executor/source/list_executor.rs:81]: make sense, why?
IIRC, before migrating to opendal, the s3 sdk returns a paginated list. I think the func name retained from then.

**Pull request merged**
